File: go/internal/logger/logger.go
package logger

import (
	"fmt"
	"log"

	"github.com/fatih/color"
)

type Level int

const (
	milestone Level = iota
	info
	warn
	softError
	hardError
)

func (l Level) color(event string) string {
	switch l {
	case milestone:
		return color.GreenString(event)
	case info:
		return color.CyanString(event)
	case warn:
		return color.YellowString(event)
	case softError:
		return color.MagentaString(event)
	case hardError:
		return color.RedString(event)
	default:
		return event
	}
}

func formatPrefix(prefix string) string {
	if prefix == "" {
		return ""
	}
	return fmt.Sprintf("[%s] ", prefix)
}

func write(prefix string, level Level, event, detail string) {
	prefix = formatPrefix(prefix)
	event = level.color(event)

	if detail == "" {
		log.Print(prefix + event)
		return
	}
	log.Print(prefix + event + " | " + detail)
}

func Milestone(event, detail string) { write("", milestone, event, detail) }
func Info(event, detail string)      { write("", info, event, detail) }
func Warn(event, detail string)      { write("", warn, event, detail) }
func SoftError(event, detail string) { write("", softError, event, detail) }
func HardError(event, detail string) { write("", hardError, event, detail) }

func Player(id uint64) string {
	return fmt.Sprintf("player=%d", id)
}

func PlayerWithAddr(id uint64, addr string) string {
	return fmt.Sprintf("player=%d addr=%s", id, addr)
}

func PlayerPair(playerId0, playerId1 uint64) string {
	return fmt.Sprintf("%s vs %s", Player(playerId0), Player(playerId1))
}

func PairPrefix(playerId0, playerId1 uint64) string {
	return fmt.Sprintf("%d vs %d", playerId0, playerId1)
}

type Logger struct {
	prefix string
}

func ForPair(playerId0, playerId1 uint64) *Logger {
	return &Logger{prefix: PairPrefix(playerId0, playerId1)}
}

func (l *Logger) Milestone(event, detail string) { write(l.prefix, milestone, event, detail) }
func (l *Logger) Info(event, detail string)      { write(l.prefix, info, event, detail) }
func (l *Logger) Warn(event, detail string)      { write(l.prefix, warn, event, detail) }
func (l *Logger) SoftError(event, detail string) { write(l.prefix, softError, event, detail) }
func (l *Logger) HardError(event, detail string) { write(l.prefix, hardError, event, detail) }

File: java/src/main/java/com/hornflakes/gorunsrestwalk/logger/Logger.java
package com.hornflakes.gorunsrestwalk.logger;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public final class Logger {

    private static final DateTimeFormatter FMT = DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm:ss");

    private static final String RESET = "\033[0m";
    private static final String GREEN = "\033[32m";
    private static final String CYAN = "\033[36m";
    private static final String YELLOW = "\033[33m";
    private static final String MAGENTA = "\033[35m";
    private static final String RED = "\033[31m";

    public static final Logger GLOBAL = new Logger("");

    private final String prefix;

    private Logger(String prefix) {
        this.prefix = prefix;
    }

    public static Logger forPair(long playerId0, long playerId1) {
        return new Logger(pairPrefix(playerId0, playerId1));
    }

    public static String player(long id) {
        return "player=" + id;
    }

    public static String playerWithAddr(long id, String addr) {
        return "player=" + id + " addr=" + addr;
    }

    public static String playerPair(long playerId0, long playerId1) {
        return player(playerId0) + " vs " + player(playerId1);
    }

    public static String pairPrefix(long playerId0, long playerId1) {
        return playerId0 + " vs " + playerId1;
    }


    public void milestone(String event, String detail) {
        write(prefix, GREEN, event, detail);
    }

    public void info(String event, String detail) {
        write(prefix, CYAN, event, detail);
    }

    public void warn(String event, String detail) {
        write(prefix, YELLOW, event, detail);
    }

    public void softError(String event, String detail) {
        write(prefix, MAGENTA, event, detail);
    }

    public void hardError(String event, String detail) {
        write(prefix, RED, event, detail);
    }

    private static void write(String prefix, String color, String event, String detail) {
        var sb = new StringBuilder();
        sb.append(LocalDateTime.now().format(FMT)).append(' ');

        if (!prefix.isEmpty()) {
            sb.append('[').append(prefix).append("] ");
        }

        sb.append(color).append(event).append(RESET);

        if (detail != null && !detail.isEmpty()) {
            sb.append(" | ").append(detail);
        }

        System.out.println(sb);
    }
}
